File: scripts/export_json.py
"""Walk every sources.toml in the repo and write one combined sources.json at the repo root.

sources.json is a derived artifact -- never hand-edit it. Any script that changes a TOML
calls main() here afterwards, so the JSON view can't go stale.

Run: python scripts/export_json.py
"""

from __future__ import annotations

import json

from common.lib import entry_to_plain_dict, load_all_docs, repo_root, today_iso


def build_json() -> tuple[dict, list[str]]:
    """Build the export document, plus whatever files wouldn't parse for main() to report."""
    docs, unreadable = load_all_docs()

    sources = []
    for path, doc in docs.items():
        # The full relative path, not just the folder name, so Tools and Tools/Free tools
        # stay distinguishable once every entry is flattened into one list.
        folder = path.parent.relative_to(repo_root()).as_posix()
        for entry in doc.get("source", []):
            record = entry_to_plain_dict(entry)
            record["folder"] = folder  # stamped last, so it reads as provenance after the entry's own fields
            sources.append(record)

    return {
        "generated_at": today_iso(),
        "count": len(sources),
        "sources": sources,
    }, unreadable


def main() -> None:
    output_path = repo_root() / "sources.json"
    data, unreadable = build_json()

    # ensure_ascii=False keeps non-ASCII readable rather than \uXXXX-escaped; the explicit
    # utf-8 matters on Windows, where the default encoding is cp1252.
    output_path.write_text(json.dumps(data, indent=2, ensure_ascii=False) + "\n", encoding="utf-8")

    print(f"Wrote {data['count']} sources to {output_path.relative_to(repo_root())}")
    for failure in unreadable:
        print(f"  skipped (unreadable): {failure}")


if __name__ == "__main__":
    main()

File: web/src/lib/export-json.ts
/** Port of scripts/export_json.py -- the flat sources.json at the repo root.
 *
 *  Output must be byte-identical to the Python's: two-space indent, non-ASCII left
 *  unescaped, one trailing newline.
 */

import { todayIso } from "./id";
import type { SourceDoc, SourceRecord, SourcesJson } from "./types";

/** Port of export_json.py:16 build_json. */
export function buildJson(docs: SourceDoc[], now: Date = new Date()): SourcesJson {
  const sources: SourceRecord[] = [];

  for (const doc of docs) {
    for (const parsed of doc.entries) {
      // `folder` is stamped last, so it reads as provenance after the entry's own fields,
      // and holds the FULL relative path so Tools and Tools/Free tools stay distinguishable.
      sources.push({ ...parsed.entry, folder: doc.folder });
    }
  }

  return {
    generated_at: todayIso(now),
    count: sources.length,
    sources,
  };
}

/** Port of export_json.py:37 main's serialisation.
 *
 *  Python writes `json.dumps(data, indent=2, ensure_ascii=False) + "\n"`. JSON.stringify
 *  with an indent of 2 produces the same layout and also leaves non-ASCII unescaped, so the
 *  only thing to add is the trailing newline.
 */
export function serializeJson(data: SourcesJson): string {
  return JSON.stringify(data, null, 2) + "\n";
}
